Presentation Only/Camera Only Streams

ata-keskin opened the pull request:
Renamed "Camera Presets" section to "Source Settings".
Added a dropdown menu to set source mode (choose between: Presentation & Camera, Presentation Only, Camera Only). This affects all past and future streams and makes sure only the specified source(s) are being used
Made sure all lecture halls settings are visible by default if there is a stream using it.
Modified how camera presets are chosen. The UI is more straightforward now. The chosen preset is marked at the beginning. The user can select the desired preset and click on save changes to save it. The previous version looked buggy in my opinion.
Modified worker code to only stream the correct sources depending on the setting chosen per course per lecture hall.
Modified /watch api to always serve PRES url or CAM url if source mode is set to presentation only (resp. camera only)
Modified client-side code to show the correct buttons on the course overview page.

Known issues:
The source mode can't be changed mid-stream. The stream has to be restarted for changes to take effect, since the workers have to start serving the sources again. I don't think we need to work on making the source made changeable mid-stream, since the users are meant to set the source mode on a per course basis.

Right now source mode selection is per (course, lecture hall) as stated above. An extension of this would be to change it to per (stream, lecture hall) which would provide more freedom to the user. However I don't know how necessary this is.

ata-keskin committed:
changed file in order to create pull request

ata-keskin committed:
Merge branch 'dev' into 226-presentation-onlycamera-only-streams

ata-keskin committed:
Renamed and changed Camera Presets section to Source Settings

instead of choosing preset by index, the user can now choose it via the preset_id, with the "value" attributes of the option tags set accordingly

ata-keskin committed:
Renamed and changed Camera Presets section to Source Settings

instead of choosing preset by index, the user can now choose it via the preset_id, with the "value" attributes of the option tags set accordingly

ata-keskin committed:
Changed camera presets updating code to just read the passed LectureHallId and SelectedPresetId and work accordingly

ata-keskin committed:
Changed camera presets updating code to just read the passed LectureHallId and SelectedPresetId and work accordingly

ata-keskin committed:
Added source preference updating code

ata-keskin committed:
Renamed camera-presets.gohtml to source-settings.gohtml

ata-keskin committed:
Made "Id" uppercase in camera_preset.go so that it looks better

ata-keskin committed:
Made "Id" uppercase in camera_preset.go so that it looks better

ata-keskin committed:
Updated the struct lhResp to include SourceMode, the id of the LectureHall and renamed SelectedIndex to SelectedPresetID

ata-keskin committed:
Updated the struct lhResp to include SourceMode, the id of the LectureHall and renamed SelectedIndex to SelectedPresetID

ata-keskin committed:
Modified worker_grpc.go:NotifyWorkers() to only create stream requests for the necessary sources, depending on the source mode specified in the source preferences of the course

atakeskinn commented:
@joschahenningsen I changed how the camera presets are chosen quite a bit. I hope I only enhanced the functionality and didn't actually break anything. The UI is much more intuitive now to say the least. I would appreciate it if you could take a look at how the Camera Presets section is looking now. I hope the idea is that the user can only choose one camera preset per lecture hall per course (I have pretty much excluded any option to do that by modfying the UI). I'm awaiting your feedback!

atakeskinn commented:
I still need to modify watch.gohtml and then hopefully it will be done :)

ata-keskin committed:
Created a new function GetSourceModeForLectureHall and replaced the usages of GetSourcePreferences with this function

This function returns the source mode for the given lecture hall id, or returns 0 if no preference for this lecture hall exists.

ata-keskin committed:
Added switch case for setting data.Version depending on source mode specified

ata-keskin committed:
Undid the change for now

ata-keskin committed:
Undid the change for now

ata-keskin committed:
Merge remote-tracking branch 'origin/226-presentation-onlycamera-only-streams' into 226-presentation-onlycamera-only-streams

# Conflicts:
#	web/watch.go

ata-keskin committed:
Changed watch.go to set clear playlist urls that are being sent.

ata-keskin committed:
Trying to check if the source mode is usable when setting source for video

ata-keskin committed:
Trying to check if the source mode is usable when setting source for video

ata-keskin committed:
Reverting the change

 and using the CanUseSource function only to decide whether to display the buttons on vod_course_list.gohtml

atakeskinn commented:
I think the implementation is complete, but I need to test it further on the test server. Feedback is appreciated!

joschahenningsen commented on worker/pb/api.pb.go:
```diff
@@ -255,7 +255,7 @@ type StreamRequest struct {
 
 	WorkerId      string                 `protobuf:"bytes,1,opt,name=WorkerId,proto3" json:"WorkerId,omitempty"`
-	SourceType    string                 `protobuf:"bytes,2,opt,name=SourceType,proto3" json:"SourceType,omitempty"`
+	SourceType    string                 `protobuf:"bytes,2,opt,name=SourceMode,proto3" json:"SourceMode,omitempty"`
```
This is a generated file and should not be edited. If you want to change the api, please edit `worker/api/api.proto` and rebuild the grpc stub (see makefile)

Why does this need to be changed though?

ata-keskin replied:
I am not sure why I changed this, I don't exactly remember changing it. I will revert it and test to make sure everything still works.

joschahenningsen commented on model/course.go:
```diff
@@ -68,6 +69,54 @@ func (c *Course) SetCameraPresetPreference(pref []CameraPresetPreference) {
 
+type SourcePreference struct {
+	LectureHallID uint `json:"lecture_hall_id"`
+	SourceMode    int  `json:"source_mode"`
```
Can you create a type for the source mode like here: https://github.com/joschahenningsen/TUM-Live/blob/15f794a088386c5fbf5f1246e14f05cb9edaa0f2/model/stream.go#L16

(this is usually how we simulate mocks in go)

ata-keskin replied:
Done

joschahenningsen commented on model/course.go:
```diff
@@ -68,6 +69,54 @@ func (c *Course) SetCameraPresetPreference(pref []CameraPresetPreference) {
+	case "COMB":
+		return mode != 1 && mode != 2
+	}
+	return false
```
In case something unexpected happens, this should default to true

ata-keskin replied:
Done.

joschahenningsen commented on model/course.go:
```diff
@@ -68,6 +69,54 @@ func (c *Course) SetCameraPresetPreference(pref []CameraPresetPreference) {
+func (c *Course) SetSourcePreference(pref []SourcePreference) {
+	pBytes, err := json.Marshal(pref)
+	if err != nil {
+		log.Println(err)
```
```suggestion
		log.WithError(err).Error("could not marshal source preference")
		return
```

joschahenningsen commented on api/worker_grpc.go:
```diff
@@ -563,73 +563,83 @@ func NotifyWorkers(daoWrapper dao.DaoWrapper) func() {
-				if sourceType != "COMB" || err != nil {
-					slot, err = daoWrapper.IngestServerDao.GetStreamSlot(server.ID)
+
+			createStreamRequest := func(sourceType string, source string) {
```
This does not need to be an inline function. Can you extract it to package level?

ata-keskin replied:
Done.

joschahenningsen commented on web/template/admin/admin_tabs/edit-course.gohtml:
```diff
@@ -25,9 +25,9 @@
 
                 <div class="bg-white dark:bg-secondary rounded shadow p-4 my-5">
-                    <h2 class="text-2 text-2xl font-semibold">Camera Presets</h2>
+                    <h2 class="text-2 text-2xl font-semibold">Source Settings</h2>
```
This could be more intuitive for lecturers:

```suggestion
                    <h2 class="text-2 text-2xl font-semibold">Lecture Hall Settings</h2>
```

joschahenningsen commented on web/template/partial/course/manage/source-settings.gohtml:
```diff
@@ -0,0 +1,56 @@
+                        @reinit.window="selected_preset = hall.presets.find(e => e.PresetID==hall.selected_preset_id); selected_preset_name = selected_preset in window ? '' : selected_preset.Name">
+                    <div class="flex justify-between">
+                        <p class="font-semibold px-1 my-auto" x-text="hall.lecture_hall_name"></p>
+                        <select class="text-black dark:bg-secondary bg-gray-100 p-2"
```
This needs some styling for dark mode.

joschahenningsen reviewed (changes requested):
Looks good already, just a few remarks :)

joschahenningsen committed:
update github actions

joschahenningsen committed:
Revert "update github actions"

This reverts commit c6f97af30fcf6360e720e6b1758ad1d793058cdf.

ata-keskin committed:
Update model/course.go

Added correct logging function.

Co-authored-by: Joscha Henningsen <44805696+joschahenningsen@users.noreply.github.com>

ata-keskin committed:
Update web/template/admin/admin_tabs/edit-course.gohtml

Renamed Camera Presets to Lecture Hall Settings

Co-authored-by: Joscha Henningsen <44805696+joschahenningsen@users.noreply.github.com>

ata-keskin committed:
Undid changes to generated source file

I don't know how these files were changed in the first place

ata-keskin committed:
Added type for SourceMode
Made sure CanUseSource returns true by default
Extracted inline function in NotifyWorkers to package level

ata-keskin committed:
Added type for SourceMode
Made sure CanUseSource returns true by default
Extracted inline function in NotifyWorkers to package level

ata-keskin committed:
Added styling for dark mode

ata-keskin committed:
Deleted typo

ata-keskin committed:
Changed type of SourceMode in lhResp as well

ata-keskin committed:
Added styling for dark mode

joschahenningsen reviewed (approved):
As discussed in the meeting, this looks great but should be merged in the semester break. Thanks for the effort! :)

MatthiasReumann committed:
Merge branch 'dev' into 226-presentation-onlycamera-only-streams

MatthiasReumann committed:
Fix tests

joschahenningsen commented on web/watch.go:
```diff
@@ -57,6 +58,23 @@ func (r mainRoutes) WatchPage(c *gin.Context) {
+
+	if tumLiveContext.Stream.LectureHallID != 0 {
+		switch data.IndexData.TUMLiveContext.Course.GetSourceModeForLectureHall(data.IndexData.TUMLiveContext.Stream.LectureHallID) {
+		// I'm not sure if this is the best solution, but it works.
```
looks good, why not :)
```suggestion
```

joschahenningsen committed:
Update web/watch.go
